### ISSUE TITLE
System-wide performance benchmarks & setting up CI-testing for them

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#58154]
This has been under discussion for a while now and it's time to finally do it. Opening this issue as an umbrella for benchmarks across the system.
FMU there seem to be 2 kinds of those in the codebase atm:
- benchmark unit tests (e.g `BenchmarkQuantityMarshalJSON`, `BenchmarkWatchHTTP`, `BenchmarkRandomStringGeneration`): Like normal unit tests, these are runnable using a single test binary - and we seem to have quite a few of these, mainly around api-machinery and helper libraries.
- benchmark integration tests (e.g `BenchmarkScheduling`, `BenchmarkSchedulingAntiAffinity`, `test/e2e_node/jenkins/benchmark`): These require to also run 1 or 2 other components (like etcd, apiserver). I can currently find these only for some components (unless I'm missing sth), mainly around node and scheduler. IMHO we should have much more of these (for e.g controller-manager, apiserver?) as they often help us with identifying performance issues in a timely, cheap and easy way.

To start with, I'm going to work towards setting up scheduler benchmarks on our CI (possibly more). However, the onus of writing useful and informative benchmarks for the respective components lies with the SIGs. I'll be happy to assist (mostly from the CI-side) as much as I can here.

cc @kubernetes/sig-scalability-misc @kubernetes/sig-testing-misc 

FYI @wojtek-t @porridge @smarterclayton @davidopp @bsalamat (folks involved in the discussions so far)

**Comment by porridge** [comment]
+1
As part of this we should make sure the results are exported and graphed
(probably on perfdash).

**Comment by porridge** [comment]
Kindly please use the "Subscribe" button on the right in the "Notifications" section rather than spam the bug with useless comments.

**Comment by ConnorDoyle** [comment]
/sig node
This topic was discussed recently in the resource management WG too

**Comment by shyamjvs** [comment]
Yes.. most probably here - http://perf-dash.k8s.io

**Comment by shyamjvs** [comment]
Ok.. So I created the [CI job](https://k8s-testgrid.appspot.com/sig-scalability-benchmarks#scheduler) running the scheduler benchmarks once every hour. It's mostly green and pretty-prints the benchmark results as an artifact in the logs. For e.g, see - https://storage.googleapis.com/kubernetes-jenkins/logs/ci-benchmark-scheduler-master/296/artifacts/BenchmarkResults.txt

What remains to be done now:
- plot those results across runs on our [performance dashboard](perf-dash.k8s.io): @porridge would most likely be looking into this
- define some kind of threshold for those benchmarks: @bsalamat @misterikkit could you come up with these?
- alert or fail the job when some threshold is crossed: I'll try taking this up

**Comment by bsalamat** [comment]
@shyamjvs 
I don't think we can find a really scientific number to be used as the threshold for the benchmarks. So, my suggestion is to take an average of say past 10 days and find standard deviation. If we don't see much variance in the numbers, we can use "average + standard deviation" as the threshold. If this sounds good to you, I can find the numbers.

**Comment by shyamjvs** [comment]
@bsalamat That sounds like a good place to start with. Could you please post the numbers below?

One more thing to consider is how reliable are those numbers. IIRC from some earlier comment of @misterikkit, those numbers have quite a bit of variance (he can confirm). If need be, we may also need to fix the machine type/architecture for running the benchmark job.

**Comment by bsalamat** [comment]
@shyamjvs I took a quick look at a few recent runs and it look to me that the variance is generally acceptable. It is a bit higher for fast tests that take <10ms per iteration, but that's expected.
I will take a more thorough look at the numbers and update here. If the variance is large then we must fix our tests or test environment. Large amount of variance in performance benchmarks means that we cannot rely on the numbers and the tests wouldn't be useful.

BTW, @misterikkit is on paternity leave. He will be back by the end of April.

**Comment by bsalamat** [comment]
Here is my suggested numbers:
```
BenchmarkScheduling/100Nodes/0Pods-8                   100    13 ms/op
BenchmarkScheduling/100Nodes/1000Pods-8                100    19 ms/op    
BenchmarkScheduling/1000Nodes/0Pods-8                  100    77 ms/op    
BenchmarkScheduling/1000Nodes/1000Pods-8               100    67 ms/op    
BenchmarkSchedulingAntiAffinity/500Nodes/250Pods-8     250   125 ms/op    
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-8    250   350 ms/op
```

Looks like the number of iterations of the first two rows is not fixed among runs. Depending the on number of iterations, the variance for those two could be high and that's not ideal. The numbers that I am suggesting here are conservative and are set so that we don't get lots of false alarms, but ideally we should run the tests with more consistent conditions to lower the variance and tighten our refrence points.

**Comment by shyamjvs** [comment]
That sounds as a good starting point. And I agree, we can adjust it later as required. Thanks @bsalamat.

**Comment by porridge** [comment]
FWIW, the benchmark graphs are now up at http://perf-dash.k8s.io/, and https://github.com/kubernetes/contrib/pull/2856 is out to bump binary version.

**Comment by shyamjvs** [comment]
That looks awesome! Thanks a lot for making those changes Marcin - we now have a nice dashboard of those results.

Btw - from those graphs, it seems like there's quite a bit of variance. So I suggest holding back for a while to get a larger sample and think about if/how we may want to define some useful alerts.

**Comment by bsalamat** [comment]
Looks great. Thanks, @porridge!

**Comment by BenTheElder** [comment]
What is left on this issue?

**Comment by wgahnagl** [comment]
/remove-sig node 
we're removing these tests so this doesn't apply to sig node anymore 
https://github.com/kubernetes/kubernetes/issues/92674
